### PR TITLE
fix(query): handle bad timezone correctly

### PR DESCRIPTION
**Pull request by mangalaman93**
in Go 1.20, time.MarshalJSON function checks for time zone strictly wrt RFC 3339 i.e. zone ∈ [0, 23]. In Dgraph, we have been simply following what time package has been doing and we need to handle the bad time zones gracefully now.

The way we handle it is that, we do not allow ingesting bad time zones going forward. Such a mutation will return an error. On the other hand, if data is already ingested, we return the result as is and ensure that such a query does not fail.

### CODE REVIEW

**Review by harshil-goel — DISMISSED** [review]

File: query/outputnode.go [hunk 1]
```diff
@@ -484,7 +484,7 @@ func (enc *encoder) AddValue(fj fastJsonNode, attr uint16, v types.Val) error {
 func (enc *encoder) AddListValue(fj fastJsonNode, attr uint16, v types.Val, list bool) error {
 	bs, err := valToBytes(v)
 	if err != nil {
-		return nil // Ignore this.
+		return fmt.Errorf("error while converting badger data to json: %v", err)
```
Comment (mangalaman93) [inline query/outputnode.go:487]: badger data -> stored data

File: query/outputnode_test.go [hunk 2]
```diff
@@ -262,3 +263,43 @@ func TestChildrenOrder(t *testing.T) {
...
+			out: "\"2018-05-30T09:30:10+30:00\""},
+		{in: time.Date(2018, 5, 30, 9, 30, 10, 0, time.FixedZone("", -99*60*60)),
+			out: "\"2018-05-30T09:30:10-99:00\""},
+		{in: time.Date(2018, 5, 30, 9, 30, 10, 0, time.FixedZone("", 99*60*60)),
+			out: "\"2018-05-30T09:30:10+99:00\""},
+	}
+
+	for _, tc := range timesToMarshal {
+		out, err := marshalTimeJson(tc.in)
+		if tc.err == "" {
```
Comment (matthewmcneely) [inline query/outputnode_test.go:297]: `tc.err` will always be "", ya? did you mean to add some failing fixtures in the timesToMarshall slice?

File: query/outputnode.go [hunk 3]
```diff
@@ -677,7 +677,7 @@ func valToBytes(v types.Val) ([]byte, error) {
 		return boolFalse, nil
 	case types.DateTimeID:
 		t := v.Value.(time.Time)
-		return t.MarshalJSON()
+		return marshalTimeJson(t)
```
Comment (anurags92) [inline query/outputnode.go:680]: add comment at the top why we need this function, maybe link golang discussion from the go repo.

File: query/outputnode.go (hunk 1 above)
Comment (mangalaman93) [reply query/outputnode.go:487]: removed the error, we are keeping the behaviour same

File: query/outputnode_test.go (hunk 2 above)
Comment (mangalaman93) [reply query/outputnode_test.go:297]: Done

File: query/outputnode.go [hunk 4]
```diff
@@ -689,6 +689,32 @@ func valToBytes(v types.Val) ([]byte, error) {
...
+	// If zone >23 or <-23, we need to handle this case ourselves.
+	// This is because, in go1.20, MarshalJSON fails for invalid zones.
+	// We, for now, call MarshalJSON for timestamp without the zone (or making it UTC zone).
+	b, err := t.Add(time.Duration(offset) * time.Second).UTC().MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+
+	// we will get a byte slice that has Z appended at the end
+	zone := offset / 60
```
[the remark this answers is not in the document]
Comment (mangalaman93) [reply query/outputnode.go:708]: Done

File: query/outputnode.go (hunk 3 above)
Comment (mangalaman93) [reply query/outputnode.go:680]: Done

**Review by matthewmcneely — DISMISSED** [review]

File: types/scalar_types.go [hunk 5]
```diff
@@ -196,6 +197,21 @@ func ValueForType(id TypeID) Val {
...
 // By default, Go parses time in UTC unless specified in the data itself.
 func ParseTime(val string) (time.Time, error) {
+	t, err := parseTimeNonStrict(val)
+	if err != nil {
+		return t, err
+	}
+
+	_, offset := t.Zone()
+	boundary := 23*60*60 + 59*60
+	if offset > boundary || offset < -1*boundary {
```
Comment (anurags92) [inline types/scalar_types.go:207]: nit: This condition is checked in another place as well. Do we want to export this in a function and call that?

**Review by anurags92 — DISMISSED** [review]

**Review by harshil-goel — DISMISSED** [review]

File: types/scalar_types.go (hunk 5 above)
Comment (mangalaman93) [reply types/scalar_types.go:207]: Done

**Review by harshil-goel — APPROVED** [review]

**Review by skrdgraph — APPROVED** [review]
cc: @rarvikar 

This looks okay. 

The plan is as follows: 
- merge this
- this PR will not break reads
- this PR will need a communication in the next release, indicating writes would error out

If customers need a fix for the `write`, we will make a patch release (as this is an anti-pattern for go-standards).

**Pull request merged**